engine/client/secretprovider/vault.go

```diff
@@ -0,0 +1,97 @@
+package secretprovider
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	vault "github.com/hashicorp/vault/api"
+	auth "github.com/hashicorp/vault/api/auth/approle"
+)
+
+var (
+	vaultClient *vault.Client
+	vaultCache  = make(map[string]map[string]any)
+)
+
+// HashiCorp Vault provider for SecretProvider
+func vaultProvider(ctx context.Context, key string) ([]byte, error) {
+	// KVv2 mount path. Default "secret"
+	mount := os.Getenv("VAULT_PATH_PREFIX")
+	if mount == "" {
+		mount = "secret"
+	}
+
+	// split key into path and field, e.g. "path/to/secret.field"
+	keyParts := strings.Split(key, ".")
+	if len(keyParts) != 2 {
+		return nil, fmt.Errorf("invalid key format: %s", key)
+	}
+	secretPath := keyParts[0]
+	secretField := keyParts[1]
+
+	// check if client is initialized
+	if vaultClient == nil {
+		err := vaultConfigureClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Seems plausible for this code to be called in parallel (if not already, then at least someday), so probably worth a `sync.Once` or similar here.

Same goes for the usage of `valueCache`

Reply:
cc @kpenfound in case you're the one working on this chunk